geometric/internal.py

```diff
@@ -3483,4 +3500,443 @@ def __init__(self, molecule, **kwargs):
+        # N3 = len(self.Prims.Internals)*3
+        # derivatives /= (val*N3)
+        # return self.w*derivatives
+           
+class EmptyCoordinates(object):
+    def __init__(self, molecule, connect=False, addcart=False, constraints=None, cvals=None, **kwargs):
+        super(EmptyCoordinates, self).__init__()
+        self.na = molecule.na
+        self.connect = connect
+        self.addcart = addcart
+        self.Internals = []
+        self.cPrims = []
+        self.cVals = []
+        self.Rotators = OrderedDict()
+        self.elem = molecule.elem
+    
+
+class ChainCoordinates(PrimitiveInternalCoordinates):
+    """
+
+    Global internal coordinate system for a chain of states.
+    Comprised of internal coordinates for each image (excluding
+    initial and final), also contains mean squared displacement
+    between pairs of images as a coordinate (experimental).
+
+    """
+    def __init__(self, molecule, connect=False, addcart=False, **kwargs):
+        super(ChainCoordinates, self).__init__(molecule, connect=connect, addcart=addcart, chain=True, **kwargs)
+        self.ic_displace = kwargs.get('ic_displace', False)
+        self.connect = connect
+        self.addcart = addcart
+        self.nim = len(molecule)
+        self.na = molecule.na
+        self.Internals = []
+        self.cPrims = []
+        self.cVals = []
+        if 'constraints' in kwargs and kwargs['constraints'] is not None:
+            raise RuntimeError('Do not use constraints with Cartesian coordinates')
+        self.elem = molecule.elem
+        self.ImageICs = [EmptyCoordinates(molecule[0], **kwargs)]
+        self.CartICs = [EmptyCoordinates(molecule[0], **kwargs)]
+        for i in range(1, len(molecule)-1):
+            if kwargs.get('cartesian', False):
+                self.ImageICs.append(CartesianCoordinates(molecule[i], **kwargs))
+            else:
+                self.ImageICs.append(PrimitiveInternalCoordinates(molecule[i], connect=connect, addcart=addcart, **kwargs))
+                self.CartICs.append(CartesianCoordinates(molecule[i], **kwargs))
+        self.ImageICs.append(EmptyCoordinates(molecule[-1], **kwargs))
+        self.CartICs.append(EmptyCoordinates(molecule[-1], **kwargs))
+        for i, imageIC in self.ICIter():
+            for ic in imageIC.Internals:
+                self.Internals.append(ImagePrim(ic, self.na, i))
+        guessw = kwargs.get('guessw', 0.1)
+        # kwargs['cartesian'] = True
+        if kwargs.get('cartesian', True):
+            #print("cartesian, w =", guessw)
+            for i, imageIC in self.ICIter():
+                self.Internals.append(RMSDisplacement(imageIC, self.na, i-1, i, head=molecule.xyzs[0]*ang2bohr, tail=molecule.xyzs[-1]*ang2bohr, w=guessw))
+            self.Internals.append(RMSDisplacement(imageIC, self.na, i+1, i, head=molecule.xyzs[0]*ang2bohr, tail=molecule.xyzs[-1]*ang2bohr, w=guessw))
+        else:
+            if self.ic_displace:
```

Comment:
I think IC displacements is only used when we use internal-coordinate NEB to get a geodesic-like reaction path without any energies. However, I don't think it works very well and I haven't used it for a few years. Have you ever used it? If not, it might be best to archive or delete these codes. This is in the overall spirit of getting rid of unused codes and simplifying codes that are hard to understand for future maintenance.

Reply:
Done. We can come back to using other IC system than Cartesian for NEB later.